File: thesis/algorithms/rust/np-dificil.rs
use std::env;
use std::fs;

fn main() {
    let args: Vec<String> = env::args().collect();
    let default_size = String::from("small");
    let size = args.get(1).unwrap_or(&default_size);
    let path = format!("datasets/{}/halting.json", size);

    let data = fs::read_to_string(&path).expect("Erro ao ler arquivo");

    // Remove colchetes e quebra em objetos individuais
    let content = data.trim();
    let content = content.trim_start_matches('[').trim_end_matches(']');
    let entries = content.split("},").collect::<Vec<&str>>();

    let mut halted = 0;
    let mut total = 0;

    for entry in entries {
        if entry.contains("\"program\"") {
            total += 1;
            if entry.contains("HALT") {
                halted += 1;
            }
        }
    }

    println!("{} de {} programas halting ({})", halted, total, size);
}

File: thesis/algorithms/java/np-dificil.java
import java.io.*;
import java.util.*;

public class NPDificil {
    public static void main(String[] args) throws IOException {
        String size = args.length > 0 ? args[0] : "small";
        String path = "../datasets/" + size + "/halting_" + size + ".json";
        BufferedReader reader = new BufferedReader(new FileReader(path));
        String line;
        int halted = 0, total = 0;
        while ((line = reader.readLine()) != null) {
            if (line.contains("program")) total++;
            if (line.contains("HALT")) halted++;
        }
        reader.close();
        System.out.println(halted + " de " + total + " programas halting (" + size + ")");
    }
}
